sweep/fee_bumper.go

```diff
@@ -76,4 +92,436 @@ type BumpResult struct {
+	)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(estimator.weight()), nil
+}
+
+// createRBFCompliantTx creates a tx that is compliant with RBF rules. It does
+// so by creating a tx, validate it using `TestMempoolAccept`, and bump its fee
+// and redo the process until the tx is valid, or return an error when non-RBF
+// related errors occur or the budget has been used up.
+func (t *TxPublisher) createRBFCompliantTx(req *BumpRequest,
+	f FeeFunction) (uint64, error) {
+
+	for {
+		// Create a new tx with the given fee rate and check its
+		// mempool acceptance.
+		tx, fee, err := t.createAndCheckTx(
+			req.Inputs, req.ChangeScript, f,
+		)
+
+		switch {
+		case err == nil:
+			// Increase the request counter.
+			//
+			// NOTE: this is the only place where we increase the
+			// counter.
+			requestID := t.requestCounter.Add(1)
+
+			// Register the record.
+			t.records.Store(requestID, &monitorRecord{
+				tx:           tx,
+				set:          req.Inputs,
+				changeScript: req.ChangeScript,
+				feeFunction:  f,
+				fee:          fee,
+			})
+
+			return requestID, nil
+
+		// If the error indicates the fees paid is not enough, we will
+		// ask the fee function to increase the fee rate and retry.
+		//
+		// TODO(yy): add this new error in btcwallet to map RBF related
+		// reject reasons:
+		// case errors.Is(err, lnwallet.ErrInsufficientFee):
+		case errors.Is(err, lnwallet.ErrMempoolFee):
+			// If the fee function tells us that we have used up
+			// the budget, we will return an error indicating this
+			// tx cannot be made. The sweeper should handle this
+			// error and try to cluster these inputs differetly.
+			err = f.IncreaseFeeRate()
+			if err != nil {
+				return 0, err
+			}
+
+		// TODO(yy): suppose there's only one bad input, we can do a
+		// binary search to find out which input is causing this error
+		// by recreating a tx using half of the inputs and check its
+		// mempool acceptance.
+		default:
+			return 0, err
+		}
+	}
+}
+
+// createAndCheckTx creates a tx based on the given inputs, change output
+// script, and the fee rate. In addition, it validates the tx's mempool
+// acceptance before returning a tx that can be published directly, along with
+// its fee.
+func (t *TxPublisher) createAndCheckTx(set InputSet, changeScript []byte,
+	f FeeFunction) (*wire.MsgTx, btcutil.Amount, error) {
+
+	// Create the sweep tx with max fee rate of 0 as the fee function
+	// guarantees the fee rate used here won't exceed the max fee rate.
+	//
+	// TODO(yy): refactor this function to not require a max fee rate.
+	tx, fee, err := createSweepTx(
+		set, nil, changeScript, uint32(t.currentHeight),
+		f.FeeRate(), 0, t.cfg.Signer,
+	)
+	if err != nil {
+		return nil, 0, fmt.Errorf("create sweep tx: %v", err)
+	}
+
+	// Sanity check the budget still covers the fee.
+	if fee > set.Budget() {
+		return nil, 0, fmt.Errorf("%w: budget=%v, fee=%v",
+			ErrNotEnoughBudget, set.Budget(), fee)
+	}
```

Comment:
Can this ever actually happen?

Reply:
I think it can happen, because we are only checking for the witness size not the overall transaction fee.
`size, _, err := pi.WitnessType().SizeUpperBound()`